### PR TITLE
[stdlib] Add equality methods to `Reference`.

### PR DESCRIPTION
**Pull request by helehex**
Does anyone know if there are any checks to take care of some equality/inequality cases at compile time? Like, by comparing the lifetimes.

### CODE REVIEW

File: stdlib/src/memory/reference.mojo [hunk 1]
```diff
@@ -332,3 +332,31 @@ struct Reference[
...
+
+        Args:
+            rhs: The value of the other pointer.
+
+        Returns:
+            True if the two pointers are equal and False otherwise.
+        """
+        return UnsafePointer(
+            __mlir_op.`lit.ref.to_pointer`(self.value)
+        ) == __mlir_op.`lit.ref.to_pointer`(rhs.value)
```
Comment (soraros) [inline stdlib/src/memory/reference.mojo:348]: Maybe wrap both with `UnsafePointer` instead of relying on implicit conversion.

Comment (helehex) [reply stdlib/src/memory/reference.mojo:348]: it looks so nice though

Comment (helehex) [reply stdlib/src/memory/reference.mojo:348]: I'll consider it. There might be a better way anyways, if not soon.

Comment (martinvuyk) [inline stdlib/src/memory/reference.mojo:348]: How about this ?
```suggestion
        alias op = __mlir_op.`lit.ref.to_pointer`
        return UnsafePointer(op(self.value)) == UnsafePointer(op(rhs.value))
```

Comment (helehex) [reply stdlib/src/memory/reference.mojo:348]: I am not strictly opposed to it

**Pull request closed without merging**